client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-recipe.ts

```diff
@@ -38,105 +54,127 @@ const useRecipe = ( siteId = 0, patterns: Pattern[], categories: Category[] ) =>
+					currentSearchParams.delete( 'header_pattern_id' );
+				} else {
+					currentSearchParams.set( 'header_pattern_id', '' + pattern.ID );
+				}
+				return currentSearchParams;
+			},
+			{ replace: true }
+		);
+	};
 
-	selectedDesignRef.current = {
-		...selectedDesign,
-		recipe: {
-			...selectedDesign?.recipe,
-			header_pattern_ids: header ? [ header.ID ] : [],
-			footer_pattern_ids: footer ? [ footer.ID ] : [],
-			pattern_ids: sections.map( ( section ) => section.ID ),
-			color_variation_title: colorVariation?.title,
-			font_variation_title: fontVariation?.title,
-		},
-	} as Design;
+	const setFooter = ( pattern: Pattern | null ) => {
+		setSearchParams(
+			( currentSearchParams ) => {
+				if ( pattern === null ) {
+					currentSearchParams.delete( 'footer_pattern_id' );
+				} else {
+					currentSearchParams.set( 'footer_pattern_id', '' + pattern.ID );
+				}
+				return currentSearchParams;
+			},
+			{ replace: true }
+		);
+	};
 
+	const incrementIndexRef = useRef( 0 );
 	const generateKey = ( pattern: Pattern ) => {
 		incrementIndexRef.current++;
 		return `${ incrementIndexRef.current }-${ pattern.ID }`;
 	};
 
-	const snapshotRecipe = useCallback( () => setSelectedDesign( selectedDesignRef.current ), [] );
+	// Adds a unique "key" prop to each section pattern,
+	// so that when the patterns are rendered in a list,
+	// they are not re-rendered unnecessarily when the list
+	// is manipulated (some elements get appended / removed).
+	const [ keyedSections, setKeyedSections ] = useState< Pattern[] >( [] );
 
-	/**
-	 * Initialize the default value from the recipe of the selected design when both patterns and categories are ready
-	 */
+	// Initializes the keys once,
+	// when the patterns and categories are finally loaded.
 	useEffect( () => {
-		if ( patterns.length === 0 || categories.length === 0 || ! selectedDesignRef.current?.recipe ) {
-			return;
-		}
-
-		const patternsById = keyBy( patterns, 'ID' );
-		const categoriesByName = keyBy( categories, 'name' );
-		const selectedHeader = patternsById[ decodePatternId( header_pattern_ids[ 0 ] ) ];
-		const selectedFooter = patternsById[ decodePatternId( footer_pattern_ids[ 0 ] ) ];
-
-		if ( selectedHeader ) {
-			setHeader( selectedHeader );
-		}
-
-		if ( selectedFooter ) {
-			setFooter( selectedFooter );
-		}
-
-		setSections(
-			pattern_ids
-				.map( ( patternId: string | number ) => patternsById[ decodePatternId( patternId ) ] )
-				.filter( Boolean )
-				.map( ( pattern: Pattern ) => {
-					const [ firstCategory ] = Object.keys( pattern.categories );
-					const category = categoriesByName[ firstCategory ];
-					return {
-						...pattern,
-						key: generateKey( pattern ),
-						category,
-					};
-				} )
-		);
+		const initialKeyedSections = sections.map( ( pattern ) => ( {
+			...pattern,
+			key: generateKey( pattern ),
+		} ) );
+
+		setKeyedSections( initialKeyedSections );
 	}, [ patterns.length, categories.length ] );
```

Comment:
This effect doesn't rely on patterns and categories anymore

Reply:
Oh I see... because we only need pattern ID to generate the key, is that what you mean?

Reply:
No, we have mapped the patterns to sections early so this effect never needs these deps. But there is a hidden logic I mentioned https://github.com/Automattic/wp-calypso/pull/80929#discussion_r1303741620

Reply:
Resolved via https://github.com/Automattic/wp-calypso/pull/80929/commits/7b514087743efe43ff916e750b603a81c3097aeb